impeller/renderer/backend/metal/surface_mtl.mm

```diff
@@ -248,13 +248,21 @@
   }
 
   if (drawable_) {
-    TRACE_EVENT0("flutter", "waitUntilScheduled");
     id<MTLCommandBuffer> command_buffer =
         ContextMTL::Cast(context.get())
             ->CreateMTLCommandBuffer("Present Waiter Command Buffer");
-    [command_buffer commit];
-    [command_buffer waitUntilScheduled];
-    [drawable_ present];
+    // If the threads have been merged, or there is a pending frame capture,
+    // then block on cmd buffer scheduling to ensure that the
+    // transaction/capture work correctly.
+    if ([[NSThread currentThread] isMainThread] ||
```

Comment:
Would this check work if a custom embedder launched the engine in a thread other than the main thread?

Reply:
I think technically this only needs to match wherever we set the layer to presentsWithTransaction. But I suspect that is only for iOS? I'll have to check how macOS desktop is doing.